### PR TITLE
Adding CUDA device-side tracing support (`--cuda_tracing`).

### LINKED ISSUE #7480: Add support for Tracy in CUDA HAL
Currently Tracy doesn't report any GPU side timing for CUDA because we haven't implemented GPU timestamp in CUDA HAL.
CUDA allows capturing GPU timestamps using CUDA events (https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__EVENT.html) however those only allow returning a differential. CUPTI tool may allow getting more fine grain information.
Once we decide how we are going to measure command buffer timing by adding timestamps at the beginning and at the end of the command buffer. It can also be used to time each dispatch independently.

After that we can hook those timestamps to tracy API the same way it is done for Vulkan:
https://github.com/google/iree/commit/99c8eaca7ffc4a776dd4d0a0f5d82ab5c12d5376#diff-3ca1acc4ea8aa3b3c0963ffa392b146aea27ac1d7a55657a5a688e4caeae1eae

- [ ] Investigate CUPTI and decide if this is the right solution to get timestamps
- [ ] Add basic timing for command buffers and add an option to enable measuring them
- [ ] Hook up the timestamps to Tracy API

### PR DESCRIPTION
**Pull request by benvanik**
This uses a pool of CUevents and trace zones inserted during stream execution and is currently capped at a reasonable amount for most models but it's something we can make dynamic in the future if needed. All zones are collected every synchronize which in today's synchronous CUDA driver is after every submission completes. In the future we'll want to more lazily process the events and only optimistically poll on submissions however since the time it takes to get the event times is so high (~10us/event) we'll probably want to build out a dedicated polling thread so that timing is not too skewed. Today several milliseconds per submission will be spent collecting timings and total execution time numbers will be inflated and users must opt-out if they want traces with accurate host timings (we could change this if we find it confusing). Additionally some of the more expensive tracing work has been moved behind the `IREE_HAL_VERBOSE_TRACING_ENABLE` flag that by default is only on in builds with assertions enabled as the user is indicating they don't need precise host timings.

Graphs will require a bit more plumbing to allow for updating the events/trace zones used during launches but since most performance work is being done with streams this is deferred. The idea is to record event nodes in the graph during command buffer recording and build a table retained on the command buffer of CUgraphNodes and the trace data - then each time a graph is launched we'd walk the table and record new events then use cuGraphExecEventRecordNodeSetEvent to update the nodes before launching. In this way we can reuse the graphs with unique timings per launch.

Example with annotations:
image
* GPU execution time: total time of command buffer (beginning of first command to end of last)
* GPU self time: all time within the total GPU execution time not attributed to dispatches (copies/fills/overhead)
* CPU command setup time: total time on host issuing CUDA API calls (beginning of first command to end of last)
* Delay to execution: **ESTIMATED** time between the first command host API call and the start of its execution on device
Note the non-trivial tracing overhead at the end of the submission!

Clicking on the command buffer zone will show all dispatches sorted by GPU time descending. Hovering over any in the list will show where they are on the timeline in the main display. It's useful to check "Group children locations" when evaluating dispatch performance though note that the same kernel may be dispatched with different grid sizes!
image

Clicking into any particular dispatch will highlight the host call that created it - in the future we can attach more information to these events at the cost of additional overhead when tracing (maybe a verbose tracing mode or something) - for now the workgroup count is only attached in builds with assertions enabled or `IREE_HAL_VERBOSE_TRACING_ENABLE` forced due to the expense:
image

Fixes #7480.

### CODE REVIEW

**Review by ScottTodd — COMMENTED** [review]
Does this fix https://github.com/iree-org/iree/issues/7480?

File: runtime/src/iree/hal/drivers/cuda/tracing.c [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+        &gpu_timestamp, &timestamp_period);
+  }
+
+  // Allocate the GPU context and pass initial calibration data.
+  if (iree_status_is_ok(status)) {
+    context->id = iree_tracing_gpu_context_allocate(
+        IREE_TRACING_GPU_CONTEXT_TYPE_VULKAN, queue_name.data, queue_name.size,
+        /*is_calibrated=*/false, cpu_timestamp, gpu_timestamp,
+        timestamp_period);
+  }
```
[the remark this answers is not in the document]
Comment (benvanik) [reply runtime/src/iree/hal/drivers/cuda/tracing.c:131]: that's what tracy calls it when gpu timestamps are equal to cpu timestamps (they've been calibrated already and tracy doesn't need to do anything) - when false tracy needs to perform calibration adjustments

File: runtime/src/iree/hal/drivers/cuda/registration/driver_module.c [hunk 2]
```diff
@@ -24,6 +24,11 @@ IREE_FLAG(bool, cuda_allow_inline_execution, false,
           "Allow command buffers to execute inline against CUDA streams when "
           "possible.");
 
+IREE_FLAG(bool, cuda_tracing, true,
+          "Enables tracing of stream events when Tracy instrumentation is "
+          "enabled. Severely impacts benchmark timings and should only be used "
+          "when analyzing dispatch timings.");
```
[the remark this answers is not in the document]
Comment (benvanik) [reply runtime/src/iree/hal/drivers/cuda/registration/driver_module.c:30]: waffled on that, but if you're building with tracing you shouldn't be running benchmarks and most people tracing cuda are likely to want to see the cuda events - the case for tracing + disabled cuda dispatch timings is just for runtime performance work and that's a much smaller set of people (me :)

File: runtime/src/iree/hal/drivers/cuda/native_executable.c [hunk 3]
```diff
@@ -57,11 +48,12 @@ iree_status_t iree_hal_cuda_native_executable_create(
 
   iree_hal_cuda_native_executable_t* executable = NULL;
 
-  // TODO: Verify the flat buffer.
+  // TODO: verify the flatbuffer.
+  // WARNING: THIS IS CURRENTLY INSECURE. This code is not checking any of the
+  // data from the flatbuffer.
   iree_CUDAExecutableDef_table_t executable_def =
       iree_CUDAExecutableDef_as_root(executable_params->executable_data.data);
```
[the remark this answers is not in the document]
Comment (benvanik) [reply runtime/src/iree/hal/drivers/cuda/native_executable.c:55]: the whole driver needs a rewrite at this point so individual issues aren't super useful. for flatbuffers we need to rework executables to contain multiple ptx blobs (same thing we need to do for vulkan) and this should be fixed as part of that (which is likely to be me) x_x

**Review by ScottTodd — APPROVED** [review]

**Pull request merged**